### PR TITLE
[Merged by Bors] - feat: add two power-of-two theorems to bitwise

### PR DESCRIPTION
**Pull request by tobiasgrosser**
These are generally useful and split out of #5920.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Nat/Bitwise.lean [hunk 1]
```diff
@@ -316,6 +319,12 @@ theorem testBit_two_pow (n m : ℕ) : testBit (2 ^ n) m = (n = m) := by
     simp [h]
 #align nat.test_bit_two_pow Nat.testBit_two_pow
 
+lemma and_two_pow {n i} : n &&& 2 ^ i = (n.testBit i).toNat * 2 ^ i := by
+  apply eq_of_testBit_eq; intro j
+  rw [mul_comm, testBit_land]
+  cases' h : n.testBit i <;> cases' (ne_or_eq i j) with h1 h1
+  <;> simp [testBit_two_pow_of_ne _, *] at * <;> assumption
```
Comment (eric-wieser) [inline Mathlib/Data/Nat/Bitwise.lean:326]: This is a non-terminal simp (more likely, one on each branch) so should be squeezed, or written a different way.

Comment (tobiasgrosser) [reply Mathlib/Data/Nat/Bitwise.lean]: I am not sure which pattern had to be avoided, but I interpreted this as adding 'only' and dropping the *. Let me know what else you had in mind.

Comment (eric-wieser) [reply Mathlib/Data/Nat/Bitwise.lean]: A "non-terminal `simp`" is a `simp` call that has no only *and* is not the last tactic in its goal. Two of your four simps in the revised version didn't need to be squeezed (via `only`), as they were terminal.

I pushed a small cleanup.

Comment (tobiasgrosser) [reply Mathlib/Data/Nat/Bitwise.lean]: Impressive. I really need to code more lean. Thank you! The current reviewing experience is outstanding.

File: Mathlib/Data/Nat/Bitwise.lean [hunk 2]
```diff
@@ -316,6 +319,14 @@ theorem testBit_two_pow (n m : ℕ) : testBit (2 ^ n) m = (n = m) := by
     simp [h]
 #align nat.test_bit_two_pow Nat.testBit_two_pow
 
+lemma and_two_pow {n i} : n &&& 2 ^ i = (n.testBit i).toNat * 2 ^ i := by
```
Comment (eric-wieser) [inline Mathlib/Data/Nat/Bitwise.lean:322]: Usually in equalities arguments are explicit:
```suggestion
lemma and_two_pow (n i : ℕ) : n &&& 2 ^ i = (n.testBit i).toNat * 2 ^ i := by
```

File: Mathlib/Data/Nat/Bitwise.lean [hunk 3]
```diff
@@ -316,6 +319,14 @@ theorem testBit_two_pow (n m : ℕ) : testBit (2 ^ n) m = (n = m) := by
     simp [h]
 #align nat.test_bit_two_pow Nat.testBit_two_pow
 
+lemma and_two_pow (n i : ℕ) : n &&& 2 ^ i = (n.testBit i).toNat * 2 ^ i := by
```
Comment (eric-wieser) [inline Mathlib/Data/Nat/Bitwise.lean:322]: Can you add `two_pow_and` for symmetry? You should prove it via commutativity, not by copying the proof.

bors d+

Thanks!

Comment (tobiasgrosser) [reply Mathlib/Data/Nat/Bitwise.lean]: Added. If the CI goes through, I will push this PR.

**Pull request closed without merging**